File: app/src/main/java/org/l3ger0j/simpledimpledraw/view/MainActivity.java
package org.l3ger0j.simpledimpledraw.view;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageButton;
import android.widget.PopupWindow;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.FileProvider;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ObservableBoolean;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.jaredrummler.android.colorpicker.ColorPickerDialog;
import com.jaredrummler.android.colorpicker.ColorPickerDialogListener;
import com.jaredrummler.android.colorpicker.ColorShape;

import org.l3ger0j.simpledimpledraw.DialogType;
import org.l3ger0j.simpledimpledraw.DrawPaint;
import org.l3ger0j.simpledimpledraw.R;
import org.l3ger0j.simpledimpledraw.ShapeBuilder;
import org.l3ger0j.simpledimpledraw.WindowType;
import org.l3ger0j.simpledimpledraw.model.DialogScreenBuilder;
import org.l3ger0j.simpledimpledraw.model.PopupWindowBuilder;
import org.l3ger0j.simpledimpledraw.model.ShakeDetector;
import org.l3ger0j.simpledimpledraw.presenter.MainContract;
import org.l3ger0j.simpledimpledraw.presenter.MainPresenter;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class MainActivity extends AppCompatActivity implements MainContract.MainActivity, ColorPickerDialogListener,
        SeekBar.OnSeekBarChangeListener {

    // region Layouts
    private PaintActivity paintActivity;
    private ShapeBuilder shapeManager;
    // endregion

    // region Buttons
    private FloatingActionButton fabSave;
    private ToggleButton toggleButton;
    // endregion

    private final MainPresenter mainPresenter = new MainPresenter(this);

    public ObservableBoolean isClickOnCloseLMenu = new ObservableBoolean();
    public ObservableBoolean isClickOnCloseRMenu = new ObservableBoolean();
    public ObservableBoolean isClickOnCloseCenterMenu = new ObservableBoolean();

    private DrawPaint drawPaint;
    private AlertDialog alertDialog;
    private TextView textView;
    private final PopupWindowBuilder popupWindowBuilder = new PopupWindowBuilder();
    private Uri uriBitmap = null;
    private String path;
    private int turnOnMove;
    private final int[] posPopupWindow = new int[2];

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // ViewBinding activity_main => ActivityMainBinding
        org.l3ger0j.simpledimpledraw.databinding.ActivityMainBinding binding =
                DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.setMainActivity(this);
        shapeManager = binding.shapeManager;
        paintActivity = binding.simpleDrawingView;

        SensorManager mSensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
        mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        ShakeDetector mShakeDetector = new ShakeDetector();
        mShakeDetector.setOnShakeListener(count -> {
            if (count == 3) {
                isClickOnCloseLMenu.set(false);
                isClickOnCloseRMenu.set(false);
                isClickOnCloseCenterMenu.set(false);
            }
        });

        drawPaint = new DrawPaint();
        paintActivity.setDrawPaint(drawPaint);

        fabSave = binding.fabSave;
        toggleButton = binding.toggleButton;

        binding.fabSetSize.setOnClickListener(mainViewOnClickList);
        binding.fabSetDrawColor.setOnClickListener(mainViewOnClickList);
        binding.fabSetBackgroundColor.setOnClickListener(mainViewOnClickList);
        binding.fabEraser.setOnClickListener(mainViewOnClickList);
        binding.fabClearCanvas.setOnClickListener(mainViewOnClickList);
        binding.fabSave.setOnClickListener(mainViewOnClickList);
        binding.fabShowMenu.setOnClickListener(mainViewOnClickList);
        binding.toggleButton.setOnClickListener(mainViewOnClickList);
        binding.undoButton.setOnClickListener(mainViewOnClickList);
        binding.redoButton.setOnClickListener(mainViewOnClickList);

        setContentView(binding.getRoot());
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        for (WindowType windowType : WindowType.values()) {
            popupWindowBuilder.createPopupWindow(windowType).dismiss();
        }
    }

    View.OnClickListener mainViewOnClickList = new View.OnClickListener() {
        @Override
        public void onClick(@NonNull View v) {
            if (v.getId() == R.id.undoButton) {
                paintActivity.undoPath();
                paintActivity.invalidateView();
            } else if (v.getId() == R.id.redoButton) {
                paintActivity.redoPath();
                paintActivity.invalidateView();
            } else if (v.getId() == R.id.fabSetSize) {
                showMinSettingDrawWindow();
            } else if (v.getId() == R.id.fabSetDrawColor) {
                createColorPickerDialog(2);
            } else if (v.getId() == R.id.fabSetBackgroundColor) {
                createColorPickerDialog(1);
            } else if (v.getId() == R.id.fabEraser) {
                if (paintActivity.isEraserOn()) {
                    paintActivity.eraseCanvas(true);
                    v.setRotation(180);
                } else {
                    paintActivity.eraseCanvas(false);
                    v.setRotation(0);
                }
            } else if (v.getId() == R.id.fabSave) {
                openCaptureMenu();
            } else if (v.getId() == R.id.fabClearCanvas) {
                paintActivity.clearCanvas();
            } else if (v.getId() == R.id.toggleButton) {
                if (shapeManager.selectShape == 1) {
                    paintActivity.drawShape(0, shapeManager);
                    shapeManager.setVisibility(View.GONE);
                    toggleButton.setVisibility(View.GONE);
                } else if (shapeManager.selectShape == 2) {
                    paintActivity.drawShape(1, shapeManager);
                    shapeManager.setVisibility(View.GONE);
                    toggleButton.setVisibility(View.GONE);
                } else if (shapeManager.selectShape == 3) {
                    paintActivity.drawShape(2, shapeManager);
                    shapeManager.setVisibility(View.GONE);
                    toggleButton.setVisibility(View.GONE);
                }
            }
        }
    };

    public void changeVisible(View v) {
        int id = v.getId();
        if (id == R.id.btnCloseLMenu) {
            isClickOnCloseLMenu.set(true);
        } else if (id == R.id.btnCloseRMenu) {
            isClickOnCloseRMenu.set(true);
        } else if (id == R.id.btnCloseCenterMenu) {
            isClickOnCloseCenterMenu.set(true);
        }
    }

    public void showAboutDialog() {
        mainPresenter.createAboutDialog().show();
    }

    public DialogInterface.OnClickListener mainDialogOnClickList (int dialogID) {
        return (dialog , which) -> {
            if (dialogID == 1) {
                // ListView listView = ((AlertDialog) dialog).getListView();
                if (which == Dialog.BUTTON_NEGATIVE) {
                    sendPic();
                }
            }
        };
    }

    // region picSave
    private void openCaptureMenu() {
        View popupView = popupWindowBuilder.createPopupWindowView(this, WindowType.CaptureMenu);
        PopupWindow popupWindow = popupWindowBuilder.createPopupWindow(WindowType.CaptureMenu);
        popupWindow.showAtLocation(fabSave, Gravity.START, 870, 140);

        popupView.findViewById(R.id.saveCanvas).setOnClickListener(v -> {
            try {
                File extStorage = getExternalFilesDir(null);
                Date date = new Date();
                SimpleDateFormat sdf = new SimpleDateFormat("yyyy_MM_dd_HH_mm", Locale.getDefault());
                String fileName = "PIC_"+sdf.format(date)+".png";
                path = extStorage.getAbsolutePath() + "/" + fileName;
                File myFile = new File(path);
                FileOutputStream fOut = new FileOutputStream(myFile);
                paintActivity.getCanvasBitmap().compress(Bitmap.CompressFormat.PNG,90,fOut);
                fOut.flush();
                fOut.close();
                uriBitmap = FileProvider.getUriForFile(
                        this,
                        "org.l3ger0j.simpledimpledraw.provider",
                        myFile
                );
                MediaStore.Images.Media.insertImage(getContentResolver(), myFile.getAbsolutePath(), myFile.getName(), myFile.getName());
                MediaScannerConnection.scanFile(getApplicationContext(), new String[]{myFile.toString()}, null, null);
            } catch (IOException e) {
                e.printStackTrace();
            }
            popupWindow.dismiss();

            Toast.makeText(this, "Save", Toast.LENGTH_LONG).show();
            DialogScreenBuilder dialogScreenBuilder = new DialogScreenBuilder();
            dialogScreenBuilder.setImageBitmap(paintActivity.getCanvasBitmap());
            dialogScreenBuilder.setOnClickListener(mainDialogOnClickList(1));
            alertDialog = DialogScreenBuilder.createAlertDialog(this , DialogType.CaptureDialog);
            alertDialog.show();
        });

        if (uriBitmap != null & path != null) {
            popupView.findViewById(R.id.shareCanvas).setOnClickListener(v -> sendPic());
        } else {
            popupView.findViewById(R.id.shareCanvas).setEnabled(false);
        }
    }

    private void sendPic() {
        Intent intent = new Intent(android.content.Intent.ACTION_SEND);
        intent.putExtra(Intent.EXTRA_STREAM , uriBitmap);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        intent.setType("image/png");
        startActivity(intent);
    }
    // endregion

    // region FAB popupWindow
    public void showMinSettingDrawWindow() {
        View popupView = popupWindowBuilder.createPopupWindowView(this, WindowType.MinimalSetting);
        PopupWindow popupWindow = popupWindowBuilder.createPopupWindow(WindowType.MinimalSetting);
        popupWindow.showAtLocation(findViewById(R.id.fabSetSize),
                Gravity.CENTER, posPopupWindow[0], posPopupWindow[1]);
        popupWindow.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));

        if (turnOnMove == 1) {
            popupView.setOnTouchListener(new View.OnTouchListener() {
                private int wx = 0;
                private int wy = 0;

                @SuppressLint("ClickableViewAccessibility")
                @Override
                public boolean onTouch(View v , MotionEvent event) {
                    switch (event.getAction()) {
                        case MotionEvent.ACTION_DOWN:
                            wx = (int) event.getX();
                            wy = (int) event.getY();
                            break;
                        case MotionEvent.ACTION_MOVE:
                            int xp = (int) event.getRawX();
                            int yp = (int) event.getRawY();
                            int sides = (xp - wx);
                            int topBot = (yp - wy);
                            popupWindow.update(sides , topBot , -1 , -1 , true);
                            posPopupWindow[0] = sides;
                            posPopupWindow[1] = topBot;
                            break;
                    }
                    return true;
                }
            });
        }

        SeekBar seekBar = popupView.findViewById(R.id.seekBar);
        seekBar.setOnSeekBarChangeListener(this);

        textView = popupView.findViewById(R.id.textView);
        textView.setText(String.valueOf(drawPaint.getStrokeWidth()));
        seekBar.setProgress((int) drawPaint.getStrokeWidth());

        ImageButton turnMoveDialog = popupView.findViewById(R.id.turnMoveDialog);
        turnMoveDialog.setOnClickListener(v -> {
            if (turnOnMove == 0) {
                turnOnMove = 1;
                popupWindow.dismiss();
                showMinSettingDrawWindow();
            } else if (turnOnMove == 1){
                turnOnMove = 0;
                popupWindow.dismiss();
                showMinSettingDrawWindow();
            }
        });
    }

    @Override
    public void onProgressChanged(@NonNull SeekBar seekBar , int progress , boolean fromUser) {
        textView.setText(String.valueOf(seekBar.getProgress()));
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    @Override
    public void onStopTrackingTouch(@NonNull SeekBar seekBar) {
        paintActivity.setStroke(seekBar.getProgress());
        shapeManager.mStrokeWidth = seekBar.getProgress();
    }
    // endregion

    private void createColorPickerDialog(int id) {
        if (id == 1) {
            ColorPickerDialog.newBuilder()
                    .setDialogId(id)
                    .setDialogTitle(R.string.color_background)
                    .setColor(paintActivity.getBackgroundColor())
                    .setDialogType(ColorPickerDialog.TYPE_PRESETS)
                    .setAllowCustom(true)
                    .setAllowPresets(true)
                    .setColorShape(ColorShape.SQUARE)
                    .show(this);
        } else if (id == 2) {
            ColorPickerDialog.newBuilder()
                    .setDialogId(id)
                    .setDialogTitle(R.string.color_drawing)
                    .setColor(drawPaint.getColor())
                    .setDialogType(ColorPickerDialog.TYPE_PRESETS)
                    .setAllowCustom(true)
                    .setAllowPresets(true)
                    .setColorShape(ColorShape.SQUARE)
                    .show(this);
        }
    }

    @Override
    public void onColorSelected(int dialogId , int color) {
        if (dialogId == 1) {
            paintActivity.setBackColor(color);
            popupWindowBuilder.setBackgroundColor(color);
        } else if (dialogId == 2) {
            paintActivity.setColor(color);
            shapeManager.mStrokeColor = color;
        }
    }

    @Override
    public void onDialogDismissed(int dialogId) {
    }
}

File: app/src/main/java/org/l3ger0j/simpledimpledraw/view/MainActivity.kt
package org.l3ger0j.simpledimpledraw.view

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.widget.ImageButton
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import android.widget.Toast
import android.widget.ToggleButton
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.jaredrummler.android.colorpicker.ColorPickerDialog
import com.jaredrummler.android.colorpicker.ColorPickerDialogListener
import com.jaredrummler.android.colorpicker.ColorShape
import org.l3ger0j.simpledimpledraw.R
import org.l3ger0j.simpledimpledraw.databinding.ActivityMainBinding
import org.l3ger0j.simpledimpledraw.utils.DrawPaint
import org.l3ger0j.simpledimpledraw.utils.ShapeBuilder
import org.l3ger0j.simpledimpledraw.view.dialogs.MainDialogFrags
import org.l3ger0j.simpledimpledraw.view.dialogs.MainDialogType
import org.l3ger0j.simpledimpledraw.view.dialogs.MainPatternDialogFrags.MainPatternDialogList
import org.l3ger0j.simpledimpledraw.view.window.PopupWindowBuilder
import org.l3ger0j.simpledimpledraw.view.window.WindowType
import org.l3ger0j.simpledimpledraw.viewModel.MainViewModel
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity(), MainPatternDialogList, ColorPickerDialogListener, OnSeekBarChangeListener {
    // region Layouts
    private var paintActivity: PaintActivity? = null
    private var shapeManager: ShapeBuilder? = null

    // endregion
    // region Buttons
    private var fabSave: FloatingActionButton? = null
    private var toggleButton: ToggleButton? = null

    // endregion
    private var drawPaint: DrawPaint? = null
    private var textView: TextView? = null
    private val popupWindowBuilder = PopupWindowBuilder()
    private var uriBitmap: Uri? = null
    private var path: String? = null
    private var turnOnMove = 0
    private val posPopupWindow = IntArray(2)
    private var mainViewModel: MainViewModel? = null
    private var mainBinding: ActivityMainBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        mainViewModel = ViewModelProvider(this)[MainViewModel::class.java]
        mainViewModel!!.mainActivityField.set(this)
        mainBinding?.mainViewModel = mainViewModel
        shapeManager = mainBinding?.shapeManager
        paintActivity = mainBinding?.simpleDrawingView
        drawPaint = DrawPaint()
        paintActivity!!.setDrawPaint(drawPaint)
        fabSave = mainBinding?.fabSave
        toggleButton = mainBinding?.toggleButton

        mainBinding?.fabSetSize?.setOnClickListener(mainViewOnClickList)
        mainBinding?.fabSetDrawColor?.setOnClickListener(mainViewOnClickList)
        mainBinding?.fabSetBackgroundColor?.setOnClickListener(mainViewOnClickList)
        mainBinding?.fabEraser?.setOnClickListener(mainViewOnClickList)
        mainBinding?.fabClearCanvas?.setOnClickListener(mainViewOnClickList)
        mainBinding?.fabSave?.setOnClickListener(mainViewOnClickList)
        mainBinding?.fabVisibility?.setOnClickListener(mainViewOnClickList)
        mainBinding?.fabShowMenu?.setOnClickListener(mainViewOnClickList)
        mainBinding?.toggleButton?.setOnClickListener(mainViewOnClickList)
        mainBinding?.undoButton?.setOnClickListener(mainViewOnClickList)
        mainBinding?.redoButton?.setOnClickListener(mainViewOnClickList)

        setContentView(mainBinding?.root)
    }

    override fun onDestroy() {
        super.onDestroy()
        for (windowType in WindowType.values()) {
            popupWindowBuilder.createPopupWindow(windowType).dismiss()
        }
    }

    @SuppressLint("NonConstantResourceId")
    var mainViewOnClickList = View.OnClickListener { v: View ->
        when (v.id) {
            R.id.undoButton -> {
                paintActivity!!.undoPath()
                paintActivity!!.invalidateView()
            }

            R.id.redoButton -> {
                paintActivity!!.redoPath()
                paintActivity!!.invalidateView()
            }

            R.id.fabSetSize -> showMinSettingDrawWindow()
            R.id.fabSetDrawColor -> createColorPickerDialog(2)
            R.id.fabSetBackgroundColor -> createColorPickerDialog(1)
            R.id.fabEraser -> if (paintActivity!!.isEraserOn()) {
                paintActivity!!.eraseCanvas(true)
                v.rotation = 180f
            } else {
                paintActivity!!.eraseCanvas(false)
                v.rotation = 0f
            }

            R.id.fabSave -> openCaptureMenu()
            R.id.fabClearCanvas -> paintActivity!!.clearCanvas()
            R.id.fabVisibility -> {}
            R.id.toggleButton -> {
                when (shapeManager!!.selectShape) {
                    1 -> {
                        paintActivity!!.drawShape(0, shapeManager!!)
                        shapeManager!!.visibility = View.GONE
                        toggleButton!!.visibility = View.GONE
                    }
                    2 -> {
                        paintActivity!!.drawShape(1, shapeManager!!)
                        shapeManager!!.visibility = View.GONE
                        toggleButton!!.visibility = View.GONE
                    }
                    3 -> {
                        paintActivity!!.drawShape(2, shapeManager!!)
                        shapeManager!!.visibility = View.GONE
                        toggleButton!!.visibility = View.GONE
                    }
                }
                mainViewModel!!.showAllMenu()
            }
        }
    }

    fun addShape(v: View) {
        when (v.id) {
            R.id.cAddCircle -> {
                shapeManager!!.selectShape = 1
                toggleButton!!.isChecked = true
                mainViewModel!!.hideAllMenu()
                shapeManager!!.visibility = View.VISIBLE
                toggleButton!!.visibility = View.VISIBLE
            }
            R.id.cAddRect -> {
                shapeManager!!.selectShape = 2
                toggleButton!!.isChecked = true
                shapeManager!!.visibility = View.VISIBLE
                mainViewModel!!.hideAllMenu()
                toggleButton!!.visibility = View.VISIBLE
            }
            R.id.cAddOval -> {
                shapeManager!!.selectShape = 3
                toggleButton!!.isChecked = true
                mainViewModel!!.hideAllMenu()
                shapeManager!!.visibility = View.VISIBLE
                toggleButton!!.visibility = View.VISIBLE
            }
        }
    }

    fun showAboutDialog() {
        val dialogFrags = MainDialogFrags()
        dialogFrags.setDialogType(MainDialogType.ABOUT_DIALOG)
        dialogFrags.show(supportFragmentManager, "aboutDialogFragment")
    }

    // region picSave
    private fun openCaptureMenu() {
        val popupView = popupWindowBuilder.createPopupWindowView(this, WindowType.CAPTURE_MENU)
        val popupWindow = popupWindowBuilder.createPopupWindow(WindowType.CAPTURE_MENU)
        popupWindow.showAtLocation(fabSave, Gravity.START, 870, 140)
        popupView?.findViewById<View>(R.id.saveCanvas)?.setOnClickListener {
            try {
                val extStorage = getExternalFilesDir(null)
                val date = Date()
                val sdf = SimpleDateFormat("yyyy_MM_dd_HH_mm", Locale.getDefault())
                val fileName = "PIC_" + sdf.format(date) + ".png"
                path = extStorage!!.absolutePath + "/" + fileName
                val myFile = File(path.toString())
                val fOut = FileOutputStream(myFile)
                paintActivity!!.canvasBitmap.compress(Bitmap.CompressFormat.PNG, 90, fOut)
                fOut.flush()
                fOut.close()
                uriBitmap = FileProvider.getUriForFile(
                        this,
                        "org.l3ger0j.simpledimpledraw.provider",
                        myFile
                )
                MediaStore.Images.Media.insertImage(contentResolver, myFile.absolutePath, myFile.name, myFile.name)
                MediaScannerConnection.scanFile(applicationContext, arrayOf(myFile.toString()), null, null)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            popupWindow.dismiss()
            Toast.makeText(this, "Save", Toast.LENGTH_LONG).show()
            val dialogFrags = MainDialogFrags()
            dialogFrags.setDialogType(MainDialogType.CAPTURE_DIALOG)
            dialogFrags.imageBitmap.set(paintActivity!!.canvasBitmap)
            dialogFrags.show(supportFragmentManager, "captureDialogFragment")
        }
        if ((uriBitmap != null) and (path != null)) {
            popupView?.findViewById<View>(R.id.shareCanvas)?.setOnClickListener { sendPic() }
        } else {
            popupView?.findViewById<View>(R.id.shareCanvas)?.isEnabled = false
        }
    }

    private fun sendPic() {
        val intent = Intent(Intent.ACTION_SEND)
        intent.putExtra(Intent.EXTRA_STREAM, uriBitmap)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intent.type = "image/png"
        startActivity(intent)
    }

    // endregion
    // region FAB popupWindow
    private fun showMinSettingDrawWindow() {
        val popupView = popupWindowBuilder.createPopupWindowView(this, WindowType.MINIMAL_SETTING)
        val popupWindow = popupWindowBuilder.createPopupWindow(WindowType.MINIMAL_SETTING)
        popupWindow.showAtLocation(findViewById(R.id.fabSetSize),
                Gravity.CENTER, posPopupWindow[0], posPopupWindow[1])
        popupWindow.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        if (turnOnMove == 1) {
            popupView?.setOnTouchListener(object : OnTouchListener {
                private var wx = 0
                private var wy = 0

                @SuppressLint("ClickableViewAccessibility")
                override fun onTouch(v: View, event: MotionEvent): Boolean {
                    when (event.action) {
                        MotionEvent.ACTION_DOWN -> {
                            wx = event.x.toInt()
                            wy = event.y.toInt()
                        }

                        MotionEvent.ACTION_MOVE -> {
                            val xp = event.rawX.toInt()
                            val yp = event.rawY.toInt()
                            val sides = xp - wx
                            val topBot = yp - wy
                            popupWindow.update(sides, topBot, -1, -1, true)
                            posPopupWindow[0] = sides
                            posPopupWindow[1] = topBot
                        }
                    }
                    return true
                }
            })
        }
        val seekBar = popupView?.findViewById<SeekBar>(R.id.seekBar)
        seekBar?.setOnSeekBarChangeListener(this)
        textView = popupView?.findViewById(R.id.textView)
        textView?.text = drawPaint!!.strokeWidth.toString()
        if (seekBar != null) {
            seekBar.progress = drawPaint!!.strokeWidth.toInt()
        }
        popupView?.findViewById<ImageButton>(R.id.turnMoveDialog)?.setOnClickListener {
            if (turnOnMove == 0) {
                turnOnMove = 1
                popupWindow.dismiss()
                showMinSettingDrawWindow()
            } else if (turnOnMove == 1) {
                turnOnMove = 0
                popupWindow.dismiss()
                showMinSettingDrawWindow()
            }
        }
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        textView!!.text = seekBar.progress.toString()
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {}
    override fun onStopTrackingTouch(seekBar: SeekBar) {
        paintActivity!!.setStroke(seekBar.progress)
        shapeManager!!.mStrokeWidth = seekBar.progress
    }

    // endregion
    private fun createColorPickerDialog(id: Int) {
        if (id == 1) {
            ColorPickerDialog.newBuilder()
                    .setDialogId(id)
                    .setDialogTitle(R.string.color_background)
                    .setColor(paintActivity!!.backgroundColor)
                    .setDialogType(ColorPickerDialog.TYPE_PRESETS)
                    .setAllowCustom(true)
                    .setAllowPresets(true)
                    .setColorShape(ColorShape.SQUARE)
                    .show(this)
        } else if (id == 2) {
            ColorPickerDialog.newBuilder()
                    .setDialogId(id)
                    .setDialogTitle(R.string.color_drawing)
                    .setColor(drawPaint!!.color)
                    .setDialogType(ColorPickerDialog.TYPE_PRESETS)
                    .setAllowCustom(true)
                    .setAllowPresets(true)
                    .setColorShape(ColorShape.SQUARE)
                    .show(this)
        }
    }

    override fun onColorSelected(dialogId: Int, color: Int) {
        if (dialogId == 1) {
            paintActivity!!.setBackColor(color)
            popupWindowBuilder.setBackgroundColor(color)
        } else if (dialogId == 2) {
            paintActivity!!.setColor(color)
            shapeManager!!.mStrokeColor = color
        }
    }

    override fun onDialogDismissed(dialogId: Int) {}
    override fun onDialogNegativeClick(mainDialogFrags: MainDialogFrags?) {
        sendPic()
    }
}